### PR TITLE
add no_namespace option

### PR DESCRIPTION
**Pull request by chris-kruining**

### CODE REVIEW

**Comment by chris-kruining** [comment]
to bring back #112

**Comment by thesayyn** [comment]
Hmm, looks delicious but tests are missing. could you add some tests for asserting the new behaviour.

Also, this looks like it only removes the top-level namespace. shouldn't it remove namespaces all the way down? 

given 

```proto
syntax = "proto3";
package foo.pkg
message Message {
  message Sub {}
}
```
i would expect 

```ts
class Message {/*omitted*/}
class Message_Sub {/*omitted*/}
```

but we get (i don't know yet cause there are no tests)

```ts
class Message {/*omitted*/}
namespace Message {
  class Sub {/*omitted*/}
}
```
whole point of this option is to remove namespaces so that users can use tools like babel

**Comment by chris-kruining** [comment]
> Hmm, looks delicious but tests are missing. could you add some tests for asserting the new behaviour.
> 
> Also, this looks like it only removes the top-level namespace. shouldn't it remove namespaces all the way down?
> 
> given
> 
> ```proto
> syntax = "proto3";
> package foo.pkg
> message Message {
>   message Sub {}
> }
> ```
> 
> i would expect
> 
> ```ts
> class Message {/*omitted*/}
> class Message_Sub {/*omitted*/}
> ```
> 
> but we get (i don't know yet cause there are no tests)
> 
> ```ts
> class Message {/*omitted*/}
> namespace Message {
>   class Sub {/*omitted*/}
> }
> ```
> 
> whole point of this option is to remove namespaces so that users can use tools like babel

I'll write those tests tomorrow

as for the namespaces, I deliberately only wanted the toplevel gone, and I think #79 does too. 
because with namespace you need to unwrap the namespace with every import you do. whilst the namespace for submessages are just a typescript workaround for subclasses

**Comment by chris-kruining** [comment]
to elaborate, what this PR wants to achieve is:
```ts
import { MyClient, MyMessage } from '@vendor/package/full/namespace/service.js';
```

instead of:
```ts
import { package } from '@vendor/package/full/namespace/service.js';
import MyClient = package.full.namespace.service.MyClient;
import MyMessage = package.full.namespace.service.MyMessage;
```

and it is not like the "root namespace" gets totally lost, as the files that are generated are in the folder of the namespace

**Comment by chris-kruining** [comment]
hmm, how would I actually go about writing a test for this, because short of adding a proto file and actually call the generator with the correct flag I have no clue as where to even start. I copied the 2 files from 'integers'(.spec.ts and .ts). but integer.ts already has no namespace for example. could you set me off in the right direction? without just implementing it all if possible, I would like to learn from this :D

**Comment by thesayyn** [comment]
perhaps adding a doc on how to add new tests, would be a good start for me but anyways. in order to add a new test using different flags, you have to create a folder under `test/_/no_namespace` and put your protos there.  then create a folder at `/test/no_namespace` and put your spec files there. of course, you will have to create the BUILD files as well. 

I would just duplicate the `/test/_/optional` and `test/optional` folders for convenience. just change the labels. to run tests just type `yarn test` and to accept diffs type `yarn update_checked_in` (this command will reflect changes you made to the plugin into generated files)

**Comment by thesayyn** [comment]
> as for the namespaces, I deliberately only wanted the toplevel gone, and I think #79 does too.

if you pay close attention to the error that is reported in the issue, you will see that babel yells at us for using namespaces.

**Comment by chris-kruining** [comment]
this is the bare minimum, but should cover what you need right? because if `no_namespace` is not working the file should export a property called `path` instead of `NoNamespace`

**Comment by chris-kruining** [comment]
O.o

File: test/_/no_namespace/BUILD.bazel [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+)
+
+ts_proto_library(
+    name = "noNamespace",
+    outs = [
+        src.replace(".proto", ".ts")
+        for src in SRCS
+    ],
+    deps = [":protos"],
+    visibility = ["//test/noNamespace:__pkg__"]
```
Comment (thesayyn) [inline test/_/no_namespace/BUILD.bazel:19]:
```suggestion
    visibility = ["//test/no_namespace:__pkg__"]
```
pattern is //{relative/path/to/BUILD}:{name}  so basically we want this target to be only visible to any target within //test/no_namespace directory.

File: test/_/no_namespace/noNamespace.proto [hunk 2]
```diff
@@ -0,0 +1,7 @@
+syntax = "proto3";
```
Comment (thesayyn) [inline test/_/no_namespace/noNamespace.proto:1]: snake case no_namespace.

File: test/no_namespace/noNamespace.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
+/**
```
Comment (thesayyn) [inline test/no_namespace/noNamespace.ts:1]: snake_case no_namespace

File: test/no_namespace/rpc.spec.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+describe('No namespace', () => {
+    it('should export a class named `NoNamespace`', async () => {
+        const message = await import('./noNamespace.js');
```
Comment (thesayyn) [inline test/no_namespace/rpc.spec.ts:3]: snake case.

**Comment by thesayyn** [comment]
do you need my help with this? I believe CI errors are descriptive enough to fix the issue.

**Comment by chris-kruining** [comment]
I have no clue

**Comment by thesayyn** [comment]
you are close. run: `yarn update_checked_in` or `npm run update_checked_in`

**Comment by chris-kruining** [comment]
can't get it to work. windows doesn't handle xargs, and my wsl ubuntu refuses to co-operate, do tou mind taking over?

**Comment by thesayyn** [comment]
> can't get it to work. windows doesn't handle xargs, and my wsl ubuntu refuses to co-operate, do tou mind taking over?

damn windows.  yeah i will take a look when i have time.

**Comment by thesayyn** [comment]
sorry, couldn't get to this one yet. a lot in my plate these days.

**Comment by michael-pomelo** [comment]
Would love for this to get merged!

Let me know if there is anything I can do to accelerate that! Would be more than glad to help out

**Comment by thesayyn** [comment]
followed by #132

**Pull request closed without merging**